pkg/controller/garbagecollector/garbagecollector.go

```diff
@@ -497,7 +497,9 @@ func monitorFor(p *Propagator, clientPool dynamic.ClientPool, resource unversion
 			},
 			UpdateFunc: func(oldObj, newObj interface{}) {
 				setObjectTypeMeta(newObj)
-				setObjectTypeMeta(oldObj)
+				// oldObj is obtained from the informer's cache, so it should
+				// have typeMeta set. Setting it again here causes data race.
+				// setObjectTypeMeta(oldObj)
```

Comment:
It's because the oldObj is get from the controller cache (https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/framework/controller.go#L238), so garbage collector's processEvent may be processing the object already. The newObj is obtained through list or watch and is fresh, so it's not causing race.